Add option to diff on and render hyperlink tag as content (show attributes inside the <a> tags)

AndriiG13 opened the pull request:
Adding functionality to ignore or monitor changes to the hyperlinks. This setting can be used to trigger alerts on changes in href attributes inside the <a> tags. 

This is set as global setting, defaulting to True (changes in links are ignored).  The global setting defaults to True, so that users don't suddenly get a diff alert on all of their tracked pages.

If setting is True, then this:
   `<a href="/link"> Some Text </a>`
renders to this:
   `Some Text`

If setting is False, then this:
    `<a href="/link"> Some Text </a>`
renders to this:
  `[ Some Text ](/link)`

The link extraction is done by using the already present `Inscriptis` package. Note that it has been updated to version 2.2 as the older version did not have this functionality.  The new version seems to have a slightly different model for extracting text, hence a few unit tests had to be adjusted. The release notes can be found here: https://github.com/weblyzard/inscriptis/releases

New unit tests have been added to test the new functionality

dgtlmoon commented:
Hi!

few things

- Rename this PR to something other people can understand, _'Add hyperlinks'_ doesn't tell me much, add hyperlinks to what? I dont get it
- _"Adding functionality to ignore or monitor changes to the hyperlinks"._ it's not clear what this does, you mean content inside the `<a ... /a>` or? can you give a use case example?
- _"This is set as global setting, defaulting to True (links are ignored)."_ new functionality should be OFF by default I think, but I'm not sure because it's not clear what this PR does
- You have test failures, did you run the tests locally?
thanks!

dgtlmoon commented on changedetectionio/fetch_site_status.py:
```diff
@@ -16,6 +17,28 @@ def __init__(self, *args, datastore, **kwargs):
         super().__init__(*args, **kwargs)
         self.datastore = datastore
 
+    def html_to_text(self, html_content: str) -> str:
```
Move this to https://github.com/dgtlmoon/changedetection.io/blob/master/changedetectionio/html_tools.py

dgtlmoon commented on changedetectionio/templates/settings.html:
```diff
@@ -82,7 +82,12 @@
+                <fieldset class="pure-group">
+                    {{ render_field(form.ignore_hyperlinks) }}
+                    <span class="pure-form-message-inline">Ignore hyperlinks, if enabled, hyperlinks will not be shown in the page content and changes in hyperlinks will not trigger alerts<br/>
+                    <i>Note:</i> Changing this will change the status of your existing watches, possibily trigger alerts etc.
```
So this is stripping out `<a... /a>` ?

dgtlmoon commented on requirements.txt:
```diff
@@ -3,7 +3,7 @@ flask~= 2.0
 validators
 timeago ~=1.0
-inscriptis ~= 1.2
+inscriptis ~= 2.2
```
Ahuh! interesting!

dgtlmoon commented:
Upgrading inscriptis over here https://github.com/dgtlmoon/changedetection.io/pull/434

I can see that just the upgrade of that library breaks the test, so the place to start is to start by making that PR pass :)

AndriiG13 commented:
Thanks for the review, I have implemented the specific change requests and added more description. 

Regarding the failing unit tests, indeed looks like just upgrading Inscriptis breaks them. It looks like since v2.0 there is a new model for handling lines. Have you had any luck with making the unit tests run on https://github.com/dgtlmoon/changedetection.io/pull/434?  

Also for context, I'm on a M1 Macbook and for some reason I'm unable to run unit tests locally unless I change my multiprocessing configuration. This is making it quite difficult to run unit tests locally, so I've been using the Github CI to debug.

Last thing, I set the new setting as True by default so that users don't suddenly get change detection alerts on all of their tracked pages. If you like, I can rephrase the setting to something like "Trigger alerts on hyperlink changes", so then the default setting can be set to False. Let me know!

dgtlmoon commented:
@AndriiG13 @timlod update your branch with https://github.com/dgtlmoon/changedetection.io master and push so the tests trigger again `2.2` update is now merged in

AndriiG13 commented:
> @AndriiG13 @timlod update your branch with https://github.com/dgtlmoon/changedetection.io master and push so the tests trigger again `2.2` update is now merged in

Nice! I updated from master and pushed, hope the tests pass now!

dgtlmoon commented:
@timlod @AndriiG13 try again on your branch but with current master from mine

AndriiG13 commented:
> @timlod @AndriiG13 try again on your branch but with current master from mine

Nice, just pushed with the changes from master

dgtlmoon commented on changedetectionio/fetch_site_status.py:
```diff
@@ -109,9 +111,13 @@ def run(self, uuid):
+                    stripped_text_from_html = \
+                        html_tools.html_to_text(
+                            html_content,
+                            ignore_hyperlinks=self.datastore.data["settings"][
```
```
ignore_hyperlinks=self.datastore.data["settings"]["application"].get("ignore_hyperlinks", False)
```

maybe is more reliable

dgtlmoon commented:
@AndriiG13 I played around with the code running and it seems to be very difficult to understand what this does exactly still

I think this is a much clearer way to explain it

```
Render anchor tag content, default disabled, when enabled renders links as <code>(link text)[http://google.com]</code>
Note: Changing this will change the status of your existing watches, possibily trigger alerts etc.
```

(note spelling error `possibily` should be `possibly`)

Please
- Change all instances of `ignore_hyperlinks` to `render_anchor_tag_content`
- Change the text there
- You may have to invert the logic so its like `ignore_hyperlinks=not render_anchor_tag_content` (feels awkward but I think this explains the impact to the user better

By default, the checkbox should be OFF but the `ignore_hyperlinks` param should be `True`

AndriiG13 commented:
Thank you for the feedback, I implemented the suggested changes!

dgtlmoon commented on changedetectionio/tests/test_ignorehyperlinks.py:
```diff
@@ -0,0 +1,155 @@
+from .util import live_server_setup
+
+import multiprocessing
+multiprocessing.set_start_method('fork')
```
What does this one do? whats the point of adding it when its not in the other tests?

AndriiG13 replied:
Have removed this. It's something I have to call to be able to run the unit test locally on my m1 mac

dgtlmoon commented on changedetectionio/tests/test_ignorehyperlinks.py:
```diff
@@ -0,0 +1,155 @@
+
+    # Cleanup everything
+    res = client.get(url_for("api_delete", uuid="all"), follow_redirects=True)
+    assert b'Deleted' in res.data
```
please add one more test case where;

1 - render hyperlinks option is OFF
2 - I load a preview/diff page that contains a hyperlink
3 - I should not see the (link)[http://foobar.com]

AndriiG13 replied:
to double check do you mean to add a test case for the "default" case, where "render_anchor_tag_content" is not passed in the post at all?

dgtlmoon replied:
that is correct

AndriiG13 replied:
added 👍

dgtlmoon commented on changedetectionio/tests/test_ignorehyperlinks.py:
```diff
@@ -0,0 +1,155 @@
+        url_for("settings_page"),
+        data={
+            "minutes_between_check": 180,
+            "render_anchor_tag_content": "false",
```
Should this be string, or boolean val?

AndriiG13 replied:
I maybe be completely off the mark, but I think it needs to be a lower case boolean, so "false". Actually any other value other than "false" evaluates to True for me and the unit test fails. 

I don't know if this is a Flask thing, but I might be missing something

dgtlmoon commented on changedetectionio/templates/settings.html:
```diff
@@ -79,10 +79,16 @@
+                    {{ render_field(form.render_anchor_tag_content) }}
+                    <span class="pure-form-message-inline">Render anchor tag content, default disabled, when enabled renders links as <code>(link text)[https://somesite.com]</code>
+                        <br/>
+                    <i>Note:</i> Changing this will change the status of your existing watches, possibly trigger alerts etc.
```
Sorry

```
Changing this could affect the content of your existing watches, possibly trigger alerts etc.
```

is maybe better

dgtlmoon commented on changedetectionio/tests/test_ignorehyperlinks.py:
```diff
@@ -151,3 +151,58 @@ def test_render_anchor_tag_content_false(client, live_server):
+    # we did not select the setting and the default behaviour is to not
+    # render anchor tag content (no new 'unviewed' class)
+    res = client.get(url_for("index"))
+    assert b"unviewed" not in res.data
```
This is not what I was thinking, as I think i wrote - you want to check that the default install does not turn on the [link](http://foobar.com) conversion

so all you need todo here is just check that, in the preview page, it does not see the rendered output ([link])

AndriiG13 replied:
aha! I see, just pushed this, let me know if it's what you had in mind

dgtlmoon commented on changedetectionio/tests/test_ignorehyperlinks.py:
```diff
@@ -8,7 +8,8 @@
     live_server_setup(live_server)
-
+import multiprocessing
+multiprocessing.set_start_method('fork')
```
remove this :)

AndriiG13 commented:
@dgtlmoon let me know if there is anything else I can do before this is ready to merge!

dgtlmoon commented on changedetectionio/fetch_site_status.py:
```diff
@@ -1,4 +1,6 @@
 import hashlib
+import urllib3
```
why is this here?

AndriiG13 replied:
must have forgotten to remove it, must have duplicated this when was resolving conflicts from master. I removed all unused imports now

dgtlmoon commented:
@AndriiG13 there is one conflict that i'm unable to solve in your branch because i dont get access to your repo, but do this

```
++<<<<<<< HEAD
 +            datastore.data['settings']['application']['render_anchor_tag_content'] = form.render_anchor_tag_content.data
 +

Remove all HEAD

 +            if form.trigger_check.data:
 +                if len(form.notification_urls.data):
 +                    n_object = {'watch_url': "Test from changedetection.io!",
 +                                'notification_urls': form.notification_urls.data,
 +                                'notification_title': form.notification_title.data,
 +                                'notification_body': form.notification_body.data,
 +                                'notification_format': form.notification_format.data,
 +                                }
 +                    notification_q.put(n_object)
 +                    flash('Test notification queued.')
 +                else:
 +                    flash('No notification URLs set, cannot send test.', 'error')

all of this is now part of the AJAX test-notification-button

++=======
keep your part
+             datastore.data['settings']['application']['real_browser_save_screenshot'] = form.real_browser_save_screenshot.data
++>>>>>>> 63e91a3d669b7b666e8fccc14974eb7c9275a88d
  

```
